### ISSUE TITLE
CDF generation - artefacts caused by calculation of CDF gradients

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#279]
At present, CDFs contain a gradient, at each point. This is actually the average of the gradient of the CDF in the bin on either side of the point.
At the ends of the CDF, the gradient is simply set equal to the gradient of the second/penultimate point. This causes artefacts in the (for example) black body spectrum near the ends of the CDF. In the attached figure one can see the artefact the right of the vertical red line which shows the end of the portion of the blackbody curve modelled by a CDF. The vertical dashed lines are photon gen band boundaries and you can see that the artefact jumps over this.

Possible solutions are to use a better estimate of the gradient - using linear extrapolation is a little better, however a tip top solution would probably be to use some kind of spline fit to the CDF...

![stellar_1em16_created_vs_emitted](https://user-images.githubusercontent.com/3329213/28622084-24b3bcb4-720b-11e7-8d77-de29cbf542dd.png)

**Comment by kslong** [comment]
@Higginbottom   Do we know what the status of this is? Do we have a .pf file that can generate the problem or see if it still exists.

**Comment by Higginbottom** [comment]
This issues has been dragged once more to the fore by Ed noting nasty wiggles in the Wien tail of his blackbody plots.

So, lets fix it. 

Here is a parameter file that can be used to recreate it...
[star.pf.txt](https://github.com/agnwinds/python/files/6384838/star.pf.txt)
and here is the spectrum of this simple 42000K star...

![star_fine_all](https://user-images.githubusercontent.com/3329213/116249043-4f10c780-a764-11eb-8451-754dd10be29c.png)
 The green lines show the photon generation band boundaries, and the red lines show ALPHAMIN=0.4, and ALPHAMAX=30.0. These are the low and high limits of where we use a full blackbody function to produce photon frequencies in the photon generation part of python. Below ALPHAMIN, we use a simple power law, above ALPHAMAX we use an exponential.

You can see some nasty artefacts on the exponential end, these are due to banding, and not this problem.

If we zoom in on the low frequency side, we clearly see the wiggle...

![star_fine_zoom_lo](https://user-images.githubusercontent.com/3329213/116249462-c34b6b00-a764-11eb-8e57-32f9c7a84d2e.png)

The wiggle is partly lining up with ALPHAMIN, and partly with a band boundary.

TO understand what is going on a quick reminder of how we compute photon frequencies from a function.

We first tabulate a PDF of the function in question. For a blackbody this is essentially emissivity vs frequency, and we tabulate this at PDFSTEPS number of points, hardwired as 10000 into cdf.c

We then make a tabulated CDF of this - for a blackbody, this starts at ALPHAMIN, and it has FUNC_CDF points, defined as 200 in python.h We basically want to know the values of alpha for which 1/200 of the emissivity occurs. This is slightly modified by the "jumps" array, which forces points in the CDF. 

The figure below shows where those CDF points are placed by the code - extracted by just hacking python to print them out...

![star_fine_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116250357-9055a700-a765-11eb-809f-ef0c7fb07220.png)

and we see that the magenta lines exactly fit with the location of the wiggle.

The reason for the wiggle is simple - in the simplest implementation of this way of doing things, we just randomly select a CDF bin for a photon to be made, and then randomly select a frequency from that small range. This would make a kind of histogram. We improve on this by computing the gradient of the CDF in each bin, and so cleverly sub-sample the frequencies in a bin given this gradient.

However, for the first bin, we don't have any information for the PDF below the lower frequency, so we can't interpolate to get a gradient. So we just use a uniform distribution - hence it is flat. The fact that there is a small break at the band edge is just because of integrating below the band ad above it  to get the total emissivity in a band. It would be *very* hard to mend this.

So, all we need to do is decide how best to get a gradient for this first CDF cell.

**Comment by Higginbottom** [comment]
OK, so here is a plot of the CDF vs alpha for the start of the blackbody CDF (on the left axis) and the differential for each point on the right axis..

![zzzzzz_cdf_zoom](https://user-images.githubusercontent.com/3329213/116262055-22fb4380-a770-11eb-8dc2-36529838421f.png)

Basically, the differential (d) of the CDF function at the ith point is given by

<img width="801" alt="image" src="https://user-images.githubusercontent.com/3329213/116262765-b7fe3c80-a770-11eb-9658-2d9f500a2dad.png">

clearly this simple implementation doesn't work for i=0 or i=FUNC_CDF so for those cases the differential calculated for the next point is used, so 

<img width="596" alt="image" src="https://user-images.githubusercontent.com/3329213/116263149-0b708a80-a771-11eb-83a2-d453732ec70a.png">

**Comment by Higginbottom** [comment]
Once a bin is selected, via the normal CDF procedure (pick a random number from 0 to 1 and translate that into a value for alpha) the code uses a quadratic root finder to discover what value of alpha corresponds to a second random number within the cell. If the gradient is the same for the upper and lower band bounds, then we end up with a uniform distribution.

One stage of increasing the complexity, is to simply extrapolate the differentials. so

<img width="524" alt="image" src="https://user-images.githubusercontent.com/3329213/116266610-8cc91c80-a773-11eb-954a-27446c1d2588.png">

certainly looks right 

![zzzzzz_cdf_zoom](https://user-images.githubusercontent.com/3329213/116266638-93f02a80-a773-11eb-9ee3-10bddab29a87.png)

Hmmmm

![star_narrow_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116267493-5d66df80-a774-11eb-8740-c611ea76180c.png)

If I move the band boundary out of the way, it looks better...

![star_narrow_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116269808-6c4e9180-a776-11eb-945c-08907feb35cf.png)

So basically our problem is that we need a better estimate of the gradient.... We could expand the CDF past where we actually want it - so we get an extra point for interpolating....

**Comment by Higginbottom** [comment]
Obviously another way of improving matters is to increase the number of points in the CDF array... going from 200 to 2000... The reason we had this quite small originally was to speed up the search for where in the CDF array a random number pointed. That has hopefully be sped up by the use of a gsl search algorithm...


![star_narrow_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116375134-e169a780-a806-11eb-9d94-6479aa76331e.png)

not sure there wont be any other fallout from this seemingly simple change tho...

We can also put in the extrapolation fix for the gradients....

![star_narrow_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116377457-1d9e0780-a809-11eb-852b-8c39d9f8a3df.png)

**Comment by Higginbottom** [comment]
Timings and effect - all with the extrapolated gradient fix.
Making 1M photons over a wide frequency range. 

FUNC_CDF=200 - time=1m23

![star_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116381961-4f18d200-a80d-11eb-9ef2-ad1038c5508f.png)

FUNC_CDF=2000 - time=1m23!!

![star_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116382359-bb93d100-a80d-11eb-96f7-066997e02c08.png)

FUNC_CDF=20000 - crashes - need to also increase PDFSTEPS from 10000 to 100000 - otherwise we run off the end of the tabulated PDF... - takes 1m15 - so clearly increasing FUNC_CDF no longer has a serious performance hit!!

![star_zoom_cdf_grid](https://user-images.githubusercontent.com/3329213/116383008-61474000-a80e-11eb-96c8-04d7df07b264.png)

No real improvement - we know that the maximum range of the BB cdf is from 0.4 to 30 - so actually going from 200 to 2000 makes some sense! 20000 is too much...

doing regression tests...

**Comment by jhmatthews** [comment]
Thanks for doing this Nick...that CIV profile blue wing (absorption to emission change) looks extremely odd at the 75 degree viewing angle.

**Comment by Higginbottom** [comment]
The emission is in the original python, it becomes absorption with the change. - I've changed the figure so the first line is old, and the second line is new. Was a bit confusing originally..

**Comment by Higginbottom** [comment]
Partially addressed by pull #846
